server/core/src/main/java/org/infinispan/server/core/backup/resources/CacheResource.java

```diff
@@ -211,32 +212,16 @@ public CacheBackupEntry next() {
                   cmd.setInternalMetadata(entry.internalMetadata);
                   return cmd;
                })
-               .flatMap(cmd -> {
-                  // Flowable does not accept null values.
-                  CompletionStage<Boolean> cs = invocationHelper.invokeAsync(cmd, 1)
-                              .thenApply(CompletableFutures.toTrueFunction());
-                  return Flowable.fromCompletionStage(cs);
-               }, batchSize)
-               .count();
-
-         return restored
-               .observeOn(Schedulers.from(blockingManager.asExecutor("restore-cache-" + cacheName)))
-               .toCompletionStage()
-               .handle((entries, t) -> {
-                  try {
-                     is.close();
-                  } catch (IOException ignore) { }
-
-                  if (t != null) {
-                     throw CompletableFutures.asCompletionException(t);
-                  }
-
-                  log.debugf("Cache %s restored %d entries", cacheName, entries);
-                  return null;
-               });
-      } catch (IOException e) {
-         throw new CacheException(e);
-      }
+               .flatMap(cmd -> RxJavaInterop.voidCompletionStageToFlowable(invocationHelper.invokeAsync(cmd, 1)), batchSize);
+      }, is -> {
+         if (is != null)
+            is.close();
+      });
+
+      return Flowable.fromPublisher(blockingManager.blockingPublisher(f))
+            .count()
```

Comment:
I am not sure if this count here uses the correct threads.

Reply:
This is fine, the key is the BlockingManager#blockingPublisher method. 

```
    * Provided a publisher that is known to block when subscribed to. Thus if the thread that subscribes in a non
    * blocking thread we will instead subscribe on a blocking thread and observe on a non blocking thread for each
    * published value.
```
So any downstream operator will be ran on a non blocking thread for all published values.